### PR TITLE
Reduce frequency of CSRF Tokens being refreshed/reset

### PR DESCRIPTION
**Pull request by tdonohue**
## References
Loosely related to https://github.com/DSpace/dspace-angular/pull/1024 , but both PRs are standalone and not dependent on one another

## Description
While testing batch uploads (see https://github.com/DSpace/dspace-angular/pull/1024), **I realized that the CSRF token is being refreshed/changed every time it is used.  This means it refreshes on every POST/PUT/PATCH/DELETE request.** While this seemed like a good idea in initial implementation, it now seems potentially problematic  / complex for batch operations.  

The reason is that during a batch upload, the CSRF token can change *between every uploaded file* (when each file is sent as a separate request). This can be difficult on the client side to achieve/manage, as any batch operations need to carefully monitor the CSRF Token. 

It also runs the risk of an (unproven) possible "race" condition: If the client sends the next request *before or while* reading the response of the last one & that response changes the CSRF token, then the "next request" may include an outdated CSRF token (causing a 403 error).

So, this PR changes the logic so that CSRF Tokens are **only reset/refreshed in the following scenarios**:
* Anytime your AuthToken changes, e.g. during Login/Logout.  This is more in line with when Spring Security defaults to changing the CSRF token...see for example their [CsrfAuthenticationStrategy](https://github.com/spring-projects/spring-security/blob/5.2.x/web/src/main/java/org/springframework/security/web/csrf/CsrfAuthenticationStrategy.java)
* If the CSRF Token was sent by the client to the server via a querystring or form parameter (as these parameters can be easily read by a third party & therefore are less secure)
* If an `InvalidCsrfTokenException` occurs, as this means the client sent an invalid CSRF token to the server. So, the server should invalidate and send a new one (allowing the client to resync its CSRF token).  This scenario is not covered in this PR, as it is already in the codebase in [DSpaceAccessDeniedHandler](https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/exception/DSpaceAccessDeniedHandler.java#L71)

## Instructions for Reviewers
* Review code & new tests to prove the new scenarios are working as expected.
   * NOTE: I also updated/added JavaDocs to `RestAuthenticationService` to make these changes easier to understand.  
* Attempt login & logout via Angular UI or Hal Browse and verify the CSRF Token changes.  Verify also that other operations (e.g. editing an Item or starting a new Submission) do not change the CSRF Token
    * This is easiest to verify from the Angular UI by opening Chrome DevTools and watching the Cookies in the "Application" tab, verifying the XSRF-TOKEN cookie updates after login & logout, but not during other operations.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/jwt/JWTTokenRestAuthenticationServiceImpl.java [hunk 1]
```diff
@@ -80,9 +86,13 @@ public void addAuthenticationDataForUser(HttpServletRequest request, HttpServlet
...
             String token = loginJWTTokenHandler.createTokenForEPerson(context, request,
                                                                  authentication.getPreviousLoginDate(), groups);
+            context.commit();
 
+            // Add newly generated auth token to the response
             addTokenToResponse(response, token, addCookie);
-            context.commit();
+
+            // Reset our CSRF token, generating a new one
+            resetCSRFToken(request, response);
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/jwt/JWTTokenRestAuthenticationServiceImpl.java:95]: @tdonohue it is not clear to me what is done here, can you help?
Are we generating a token on line 87 and immediately generate a new one in line 95? if this is the case, is not the token in the cookie different from the one send via header in line 95?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java [hunk 2]
```diff
@@ -368,6 +374,10 @@ public void testRefreshToken() throws Exception {
 
         String newToken = getClient(token).perform(post("/api/authn/login"))
                                           .andExpect(status().isOk())
+                                          // An auth token refresh should also refresh the CSRF token
+                                          // (which should be returned in both a cookie & header).
+                                          .andExpect(cookie().exists("DSPACE-XSRF-COOKIE"))
+                                          .andExpect(header().exists("DSPACE-XSRF-TOKEN"))
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:380]: it would be nice if we can check that the token is the same in both place

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java [hunk 3]
```diff
@@ -949,11 +976,29 @@ public void testShortLivedToken() throws Exception {
...
+    public void testShortLivedTokenWithCSRFSentViaParam() throws Exception {
+        String token = getAuthToken(eperson.getEmail(), password);
+
+        // Same request as prior method, but this time we are sending the CSRF token as a querystring param.
+        // NOTE: getClient() method defaults to sending CSRF tokens as Headers, so we are overriding its behavior here
+        getClient(token).perform(post("/api/authn/shortlivedtokens").with(csrf()))
+            // BECAUSE we sent the CSRF token on querystring, it should be regenerated & a new token
+            // is sent back (in cookie and header).
+            .andExpect(cookie().exists("DSPACE-XSRF-COOKIE"))
+            .andExpect(header().exists("DSPACE-XSRF-TOKEN"));
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:998]: it would be better to verify that the token value is really changed

**Review by abollini — COMMENTED** [review]
hi @tdonohue I have put a question inline, can you help me to understand the code?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/jwt/JWTTokenRestAuthenticationServiceImpl.java (hunk 1 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/security/jwt/JWTTokenRestAuthenticationServiceImpl.java:95]: @abollini : You've confused the two tokens here.  Line 92 (`addTokenToResponse`) is adding the *Auth* Token (JWT) to the response.  Line 95 (`resetCSRFToken`) is resetting the *CSRF* Token.  So, these are two entirely separate tokens.  The logic here is that anytime you create a *new* auth token (JWT) we should immediately reset the CSRF Token....this ensures that the CSRF token changes anytime you login.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java (hunk 2 above)
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:380]: Good point. I can add a line to check that the cookie() matches the header().  It always should, but it's worth checking.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java (hunk 3 above)
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:998]: @abollini : Unfortunately, I've not found a way in these tests to verify the "before" and "after" value of the CSRF token...the tests themselves generate the CSRF automatically by just calling `csrf()` (see line 994).  So, the reason I simply check the existence of the cookie & header is that those are ONLY returned in the response when the token changes.  If the token doesn't change, then neither the cookie nor the header exists (see the previous test on lines 982-983).

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java (hunk 2 above)
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:380]: @abollini : On second thought, this is much harder than it seems :(  There's no easy way in Spring to check that that value returned in a Cookie is the same as the one in a Header.  

Also, we've modeled the behavior of Spring Security's `CsrfAuthenticationStrategy` code, which technically sets the Cookie *twice* in one response...first to an empty string ("") to clear it out, then to the new value (see https://github.com/spring-projects/spring-security/blob/master/web/src/main/java/org/springframework/security/web/csrf/CsrfAuthenticationStrategy.java#L58-L60).  So, the response technically looks like:
```
Set-Cookie: DSPACE-XSRF-COOKIE=; [other settings]
Set-Cookie: DSPACE-XSRF-COOKIE=[new-token]; [other settings]
DSPACE-XSRF-TOKEN: [new-token]
```

Spring's `cookie()` matcher seems to assume you only have one Cookie of a name, and there's no way to tell `cookie().value()` to match the *second* value and not the *first* one (which will always be an empty string when a token change occurs).

In other words, I think this test might be as good as we can get without enhancing the test framework overall (which would take more time than we currently have).  We already have proven (in other tests) that this Cookie & Header are only returned when we expect a CSRF token change to occur.

**Comment by tdonohue** [comment]
@abollini : I've responded to your questions above.  As noted in those responses, I don't think there's any code changes necessary for this PR, but let me know if you disagree.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java (hunk 2 above)
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:380]: The best I've come up with in terms of *verifying* the CSRF Token, is something like this:
```
.andExpect(cookie().exists("DSPACE-XSRF-COOKIE"))
.andExpect(header().exists("DSPACE-XSRF-TOKEN"))
// First cookie will always be empty. Second has the token & there's no way to match it easily
.andExpect(cookie().value("DSPACE-XSRF-COOKIE", ""))
// CSRF Tokens generated by Spring Security are UUIDs
.andExpect(header().string("DSPACE-XSRF-TOKEN", matchesPattern(REGEX_UUID)));
```

This isn't what you've asked for, but it's slightly better than just checking for existence.  If you feel this is better, I can add it to the tests here.

Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:380]: thanks @tdonohue this is absolutely ok as it tests exactly what you (we) expect to happen. I'm not sure that it is a good idea to add an empty cookie before as the new one should already update any previous information on the browser side but as I was not able to find any specs that prohibits to do that and there are no evidence of side effect I'm ok with that

**Comment by abollini** [comment]
I'm ok with the improvement that you have suggested, +1 to merge

**Review by abollini — APPROVED** [review]

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java (hunk 2 above)
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:380]: Just in case we are wondering why the empty cookie approach is being used.  Again, I chose that approach as it's the same approach used by Spring Security in their `CsrfAuthenticationStrategy`:
1. First, they [set the CSRF token to `null`](https://github.com/DSpace/dspace-angular/issues/1069), which creates an empty cookie first
2. Then, they [set the new Token](https://github.com/spring-projects/spring-security/blob/5.4.x/web/src/main/java/org/springframework/security/web/csrf/CsrfAuthenticationStrategy.java#L60), which creates a second cookie with the token in it.

I'll make sure to document this as well in the comments of the test itself (which I'll create shortly)
